File: app/Catalogo/GameMod/services/external-tool-installer.ts
import fs from "node:fs";
import path from "node:path";
import os from "node:os";
import { MakaiRPC } from "@mods-manager/services/makai-rpc";
import { logger } from "@main/services";
import type { ExternalToolDef } from "@games/_shared/types";

function getToolsBaseDir(): string {
  return path.join(os.homedir(), ".config", "makai-forger", "tools");
}

export function getToolInstallDir(gameId: string, toolName: string): string {
  const safeName = toolName.replace(/[\/\\]/g, "_");
  return path.join(getToolsBaseDir(), gameId, safeName);
}

export function getGameToolsDir(gameId: string): string {
  return path.join(getToolsBaseDir(), gameId);
}

export function isToolInstalled(gameId: string, tool: ExternalToolDef): boolean {
  const toolDir = getToolInstallDir(gameId, tool.name);
  if (tool.detector?.folder) {
    return fs.existsSync(path.join(toolDir, tool.detector.folder));
  }
  if (tool.detector?.file) {
    return fs.existsSync(path.join(toolDir, tool.detector.file));
  }
  return fs.existsSync(path.join(toolDir, tool.exeName));
}

export function resolveToolPath(gameId: string, tool: ExternalToolDef): string | null {
  const toolDir = getToolInstallDir(gameId, tool.name);
  if (tool.detector?.file) {
    const p = path.join(toolDir, tool.detector.file);
    if (fs.existsSync(p)) return p;
  }
  if (tool.detector?.folder) {
    const p = path.join(toolDir, tool.detector.folder, tool.exeName);
    if (fs.existsSync(p)) return p;
  }
  const p = path.join(toolDir, tool.exeName);
  if (fs.existsSync(p)) return p;
  return null;
}

export async function installTool(
  gameId: string,
  tool: ExternalToolDef,
  send?: (step: string, msg: string, type: string) => void,
): Promise<boolean> {
  if (!tool.downloadUrl) return false;

  const toolDir = getToolInstallDir(gameId, tool.name);
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), `tool-${tool.name.replace(/\s+/g, "_")}-`));

  try {
    const urlLower = tool.downloadUrl.toLowerCase();
    const ext = urlLower.endsWith(".7z") ? "7z" : urlLower.endsWith(".zip") ? "zip" : "zip";
    const archivePath = path.join(tmpDir, `tool.${ext}`);

    send?.("tools", `Baixando ${tool.name}...`, "working");
    await MakaiRPC.call("download_file", { url: tool.downloadUrl, dest: archivePath });

    if (!fs.existsSync(archivePath) || fs.statSync(archivePath).size === 0) {
      throw new Error("Download falhou: arquivo vazio");
    }

    send?.("tools", `Extraindo ${tool.name}...`, "working");
    const extractDir = path.join(tmpDir, "extracted");
    fs.mkdirSync(extractDir, { recursive: true });

    await MakaiRPC.call("extract_archive", { archive: archivePath, dest: extractDir });

    let sourceDir = extractDir;
    if (tool.innerFolder) {
      const innerPath = path.join(extractDir, tool.innerFolder);
      if (fs.existsSync(innerPath)) {
        sourceDir = innerPath;
      }
    } else {
      const entries = fs.readdirSync(extractDir, { withFileTypes: true });
      const dirs = entries.filter(e => e.isDirectory());
      const files = entries.filter(e => e.isFile());
      if (dirs.length === 1 && files.length === 0) {
        sourceDir = path.join(extractDir, dirs[0].name);
      }
    }

    send?.("tools", `Instalando ${tool.name}...`, "working");
    fs.mkdirSync(toolDir, { recursive: true });
    copyRecursive(sourceDir, toolDir);

    const installed = isToolInstalled(gameId, tool);
    if (installed) {
      send?.("tools", `${tool.name} instalado em ${toolDir}`, "done");
    } else {
      send?.("tools", `${tool.name} instalado mas exe não encontrado`, "done");
    }

    return installed;
  } catch (err) {
    const msg = String(err).slice(0, 200);
    logger.error(`[Tool] ${tool.name} install failed: ${msg}`);
    send?.("tools", `Falha ao instalar ${tool.name}: ${msg}`, "error");
    return false;
  } finally {
    try { fs.rmSync(tmpDir, { recursive: true, force: true }); } catch { /* skip */ }
  }
}

export async function ensureExternalTools(
  gameId: string,
  tools: ExternalToolDef[],
  send?: (step: string, msg: string, type: string) => void,
): Promise<{ installed: string[]; skipped: string[]; failed: string[] }> {
  const installed: string[] = [];
  const skipped: string[] = [];
  const failed: string[] = [];

  const downloadable = tools.filter(t => t.downloadUrl);
  if (downloadable.length === 0) {
    return { installed, skipped, failed };
  }

  for (const tool of downloadable) {
    if (isToolInstalled(gameId, tool)) {
      send?.("tools", `${tool.name} já instalado`, "done");
      skipped.push(tool.name);
      continue;
    }

    const ok = await installTool(gameId, tool, send);
    if (ok) {
      installed.push(tool.name);
    } else {
      failed.push(tool.name);
    }
  }

  return { installed, skipped, failed };
}

function copyRecursive(src: string, dst: string): void {
  let entries: fs.Dirent[];
  try { entries = fs.readdirSync(src, { withFileTypes: true }); }
  catch { return; }

  for (const entry of entries) {
    const srcPath = path.join(src, entry.name);
    const dstPath = path.join(dst, entry.name);

    if (entry.isDirectory()) {
      fs.mkdirSync(dstPath, { recursive: true });
      copyRecursive(srcPath, dstPath);
    } else if (entry.isFile()) {
      fs.mkdirSync(path.dirname(dstPath), { recursive: true });
      fs.copyFileSync(srcPath, dstPath);
      try { fs.chmodSync(dstPath, 0o755); } catch { /* skip */ }
    }
  }
}

File: app/Catalogo/GameMod/core/engine/external_tool_installer.py
"""
core/engine/external_tool_installer.py — Instalação de external tools (SSEEdit, LOOT, WolvenKit, etc.)

ExternalToolDef:
  name        - Nome exibido
  exeName     - Nome do executável
  downloadUrl - URL direta (zip/7z) — opcional
  detector    - {file?, folder?} para checar se já instalado
  innerFolder - Pasta dentro do zip com os arquivos
  useProton   - Se true, lança via Proton
  args        - Args extras ao lançar
"""

import json
import os
import shutil
import sys
import tempfile
import urllib.request

_BASE_TOOLS_DIR = os.path.expanduser("~/.config/makai-forger/tools")

# ─── Definições ──────────────────────────────────────────────────

_BETHESA_TOOLS = [
    {"name": "SSEEdit", "exeName": "SSEEdit.exe", "searchPaths": ["."]},
    {"name": "FNIS", "exeName": "FNIS.exe", "searchPaths": ["."]},
    {"name": "BodySlide", "exeName": "BodySlide.exe", "searchPaths": ["."]},
    {"name": "Outfit Studio", "exeName": "OutfitStudio.exe", "searchPaths": ["."]},
    {"name": "LOOT", "exeName": "LOOT.exe", "searchPaths": ["."]},
    {"name": "Wrye Bash", "exeName": "Wrye Bash.exe", "searchPaths": ["."]},
    {"name": "Creation Kit", "exeName": "CreationKit.exe", "searchPaths": ["."]},
    {"name": "zEdit", "exeName": "zEdit.exe", "searchPaths": ["."]},
    {"name": "Cathedral Assets Optimizer", "exeName": "CAO.exe", "searchPaths": ["."]},
    {"name": "Nemesis", "exeName": "Nemesis Unlimited Behavior Engine.exe", "searchPaths": ["."]},
    {"name": "BethINI", "exeName": "BethINI.exe", "searchPaths": ["."]},
]

_SKYRIM_TOOLS = [
    {"name": "SSEEdit", "exeName": "xTESEdit64.exe", "searchPaths": ["."],
     "downloadUrl": "https://github.com/TES5Edit/TES5Edit/releases/latest/download/xEdit.4.1.5f.7z",
     "detector": {"file": "xTESEdit64.exe"}},
    {"name": "LOOT", "exeName": "LOOT.exe", "searchPaths": ["."],
     "downloadUrl": "https://github.com/loot/loot/releases/latest/download/loot_0.29.1-win64.7z",
     "detector": {"file": "LOOT.exe"}},
    {"name": "Wrye Bash", "exeName": "Wrye Bash.exe", "searchPaths": ["."],
     "downloadUrl": "https://github.com/Wrye-Bash/Wrye-Bash/releases/latest/download/Wrye.Bash.314.-.Standalone.Executable.7z",
     "innerFolder": "Mopy", "detector": {"file": "Wrye Bash.exe"}},
    {"name": "zEdit", "exeName": "zEdit.exe", "searchPaths": ["."],
     "downloadUrl": "https://github.com/z-edit/zedit/releases/download/0.6.7/zEdit_v0.6.7_-_Portable_x64.7z",
     "detector": {"file": "zEdit.exe"}},
    {"name": "ESLifier", "exeName": "ESLifier.exe", "searchPaths": ["."],
     "downloadUrl": "https://github.com/MaskPlague/ESLifier/releases/download/v0.15.3/ESLifier.zip",
     "detector": {"file": "ESLifier.exe"}},
    {"name": "Pandora Behavior Engine+", "exeName": "Pandora Behaviour Engine+.exe", "searchPaths": ["."],
     "downloadUrl": "https://github.com/Monitor221hz/Pandora-Behaviour-Engine-Plus/releases/latest/download/Pandora_Behaviour_Engine_v4.3.1-beta_win-x64_net.zip",
     "detector": {"file": "Pandora Behaviour Engine+.exe"}},
    {"name": "FNIS", "exeName": "FNIS.exe", "searchPaths": ["."]},
    {"name": "BodySlide", "exeName": "BodySlide.exe", "searchPaths": ["."]},
    {"name": "Outfit Studio", "exeName": "OutfitStudio.exe", "searchPaths": ["."]},
    {"name": "Creation Kit", "exeName": "CreationKit.exe", "searchPaths": ["."]},
    {"name": "Cathedral Assets Optimizer", "exeName": "CAO.exe", "searchPaths": ["."]},
    {"name": "Nemesis", "exeName": "Nemesis Unlimited Behavior Engine.exe", "searchPaths": ["."]},
    {"name": "BethINI", "exeName": "BethINI.exe", "searchPaths": ["."]},
    {"name": "DynDOLOD", "exeName": "DynDOLODx64.exe", "searchPaths": ["."]},
    {"name": "TexGen", "exeName": "TexGenx64.exe", "searchPaths": ["."]},
    {"name": "xLODGen", "exeName": "xLODGenx64.exe", "searchPaths": ["."]},
    {"name": "VRAMr", "exeName": "VRAMr.exe", "searchPaths": ["."]},
    {"name": "BENDr", "exeName": "BENDr.exe", "searchPaths": ["."]},
    {"name": "ParallaxR", "exeName": "ParallaxR.exe", "searchPaths": ["."]},
]

_MORROWIND_TOOLS = [
    {"name": "TES3Edit", "exeName": "TES3Edit.exe", "searchPaths": ["."],
     "downloadUrl": "https://github.com/TES5Edit/TES5Edit/releases/latest/download/xEdit.4.1.5f.7z",
     "detector": {"file": "TES3Edit.exe"}},
    {"name": "LOOT", "exeName": "LOOT.exe", "searchPaths": ["."],
     "downloadUrl": "https://github.com/loot/loot/releases/latest/download/loot_0.29.1-win64.7z",
     "detector": {"file": "LOOT.exe"}},
    {"name": "Wrye Mash", "exeName": "Wrye Mash.exe", "searchPaths": ["."]},
    {"name": "Construction Set", "exeName": "TESConstructionSet.exe", "searchPaths": ["."]},
    {"name": "MWSE", "exeName": "mwse_loader.exe", "searchPaths": ["."],
     "downloadUrl": "https://github.com/MWSE/MWSE/releases/download/2.1/MWSE-2.1.7z",
     "detector": {"file": "mwse_loader.exe"}},
    {"name": "MGE XE", "exeName": "MGEXEgui.exe", "searchPaths": ["."]},
]

_GAME_TOOLS: dict[str, list[dict]] = {
    "skyrim": _SKYRIM_TOOLS,
    "skyrim_se": _SKYRIM_TOOLS,
    "enderal": _SKYRIM_TOOLS,
    "fallout3": _BETHESA_TOOLS,
    "fallout4": _BETHESA_TOOLS,
    "fallout4_vr": _BETHESA_TOOLS,
    "falloutnv": _BETHESA_TOOLS,
    "oblivion": _BETHESA_TOOLS,
    "starfield": _BETHESA_TOOLS,
    "morrowind": _MORROWIND_TOOLS,
    "cyberpunk2077": [
        {"name": "WolvenKit", "exeName": "WolvenKit.exe", "searchPaths": ["."],
         "downloadUrl": "https://github.com/WolvenKit/WolvenKit/releases/download/8.19.0/WolvenKit-8.19.0.zip",
         "detector": {"file": "WolvenKit.exe"}, "useProton": True},
        {"name": "ArchiveXL", "exeName": "", "searchPaths": []},
        {"name": "TweakXL", "exeName": "", "searchPaths": []},
        {"name": "Codeware", "exeName": "", "searchPaths": []},
    ],
    "larian": [
        {"name": "BG3 Mod Manager", "exeName": "bg3mm.exe", "searchPaths": ["."]},
    ],
    "witcher3": [
        {"name": "Script Merger", "exeName": "ScriptMerger.exe", "searchPaths": ["."]},
        {"name": "Witcher 3 Mod Limit Fix", "exeName": "", "searchPaths": []},
    ],
}


def get_tools(game_id: str) -> list[dict]:
    return _GAME_TOOLS.get(game_id, [])


def is_tool_installed(game_id: str, tool: dict) -> bool:
    tool_dir = _get_tool_install_dir(game_id, tool["name"])
    detector = tool.get("detector") or {}
    folder = detector.get("folder")
    file = detector.get("file")
    if folder:
        return os.path.isdir(os.path.join(tool_dir, folder))
    if file:
        return os.path.isfile(os.path.join(tool_dir, file))
    exe = tool.get("exeName")
    if exe:
        return os.path.isfile(os.path.join(tool_dir, exe))
    return os.path.isdir(tool_dir)


def resolve_tool_path(game_id: str, tool: dict) -> str | None:
    tool_dir = _get_tool_install_dir(game_id, tool["name"])
    detector = tool.get("detector") or {}
    if detector.get("file"):
        p = os.path.join(tool_dir, detector["file"])
        if os.path.isfile(p):
            return p
    if detector.get("folder"):
        p = os.path.join(tool_dir, detector["folder"], tool.get("exeName", ""))
        if os.path.isfile(p):
            return p
    exe = tool.get("exeName")
    if exe:
        p = os.path.join(tool_dir, exe)
        if os.path.isfile(p):
            return p
    return None


def _get_tool_install_dir(game_id: str, tool_name: str) -> str:
    safe = tool_name.replace("/", "_").replace("\\", "_")
    return os.path.join(_BASE_TOOLS_DIR, game_id, safe)


def _extract_archive(archive_path: str, dest_dir: str) -> bool:
    import zipfile
    try:
        with zipfile.ZipFile(archive_path) as z:
            z.extractall(dest_dir)
        return True
    except Exception:
        pass
    try:
        import py7zr
        with py7zr.SevenZipFile(archive_path, mode="r") as z:
            z.extractall(path=dest_dir)
        return True
    except Exception:
        pass
    try:
        import subprocess
        r = subprocess.run(["7z", "x", archive_path, f"-o{dest_dir}", "-y"],
                           capture_output=True, timeout=30)
        return r.returncode == 0
    except Exception:
        pass
    return False


def _copy_recursive(src: str, dst: str):
    for entry in os.listdir(src):
        s = os.path.join(src, entry)
        d = os.path.join(dst, entry)
        if os.path.isdir(s):
            os.makedirs(d, exist_ok=True)
            _copy_recursive(s, d)
        elif os.path.isfile(s):
            os.makedirs(os.path.dirname(d), exist_ok=True)
            shutil.copy2(s, d)
            try:
                os.chmod(d, 0o755)
            except OSError:
                pass


def _get_inner_dir(extract_dir: str, tool: dict) -> str:
    inner = tool.get("innerFolder")
    if inner:
        p = os.path.join(extract_dir, inner)
        if os.path.isdir(p):
            return p
    entries = os.listdir(extract_dir)
    dirs = [e for e in entries if os.path.isdir(os.path.join(extract_dir, e))]
    files = [e for e in entries if os.path.isfile(os.path.join(extract_dir, e))]
    if len(dirs) == 1 and len(files) == 0:
        return os.path.join(extract_dir, dirs[0])
    return extract_dir


def install_tool(game_id: str, tool: dict) -> bool:
    url = tool.get("downloadUrl")
    if not url:
        return False

    tool_dir = _get_tool_install_dir(game_id, tool["name"])

    with tempfile.TemporaryDirectory(prefix=f"tool-{tool['name']}-") as tmpdir:
        ext = ".7z" if url.lower().endswith(".7z") else ".zip"
        archive_path = os.path.join(tmpdir, f"tool{ext}")

        _emit("log", level="info", message=f"Baixando {tool['name']}...")
        try:
            urllib.request.urlretrieve(url, archive_path)
        except Exception as e:
            _emit("log", level="warn", message=f"Download {tool['name']} falhou: {e}")
            return False

        if not os.path.isfile(archive_path) or os.path.getsize(archive_path) == 0:
            _emit("log", level="warn", message=f"Download {tool['name']} vazio")
            return False

        _emit("log", level="info", message=f"Extraindo {tool['name']}...")
        extract_dir = os.path.join(tmpdir, "extracted")
        os.makedirs(extract_dir, exist_ok=True)
        if not _extract_archive(archive_path, extract_dir):
            _emit("log", level="warn", message=f"Extração {tool['name']} falhou")
            return False

        source_dir = _get_inner_dir(extract_dir, tool)

        _emit("log", level="info", message=f"Instalando {tool['name']}...")
        os.makedirs(tool_dir, exist_ok=True)
        _copy_recursive(source_dir, tool_dir)

        installed = is_tool_installed(game_id, tool)
        if installed:
            _emit("log", level="info", message=f"{tool['name']} instalado em {tool_dir}")
        else:
            _emit("log", level="warn", message=f"{tool['name']}: exe não encontrado")
        return installed


def ensure_tools(game_id: str) -> dict:
    tools = get_tools(game_id)
    installed = []
    skipped = []
    failed = []

    downloadable = [t for t in tools if t.get("downloadUrl")]
    for tool in downloadable:
        if is_tool_installed(game_id, tool):
            _emit("log", level="info", message=f"{tool['name']} já instalado")
            skipped.append(tool["name"])
            continue
        ok = install_tool(game_id, tool)
        if ok:
            installed.append(tool["name"])
        else:
            failed.append(tool["name"])

    return {"installed": installed, "skipped": skipped, "failed": failed}


def _emit(level: str, message: str, **extra):
    try:
        payload = {"event": "log", "level": level, "message": message, **extra}
        sys.stdout.write(json.dumps(payload, separators=(",", ":")) + "\n")
        sys.stdout.flush()
    except Exception:
        pass
